chore: migrate azure-sdk-for-go/containerregistry to the latest release

shahramk64 opened the pull request:
# Description

## What this PR does / why we need it:
auth provider currently uses an old preview version of azure sdk for go. With the latest release of the sdk, the necessary API to exchange the AAD access token for an ACR refresh token is exposed and we can migrate to this latest release.

## Which issue(s) this PR fixes *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #959 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Helm Chart Change (any edit/addition/update that is necessary for changes merged to the `main` branch)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Here is a link to the successful AKS run:
https://github.com/shahramk64/forked_ratify/actions/runs/11024417178/job/30676580847?pr=2

# Checklist:

- [ ] Does the affected code have corresponding tests?
- [ ] Are the changes documented, not just with inline documentation, but also with conceptual documentation such as an overview of a new feature, or task-based documentation like a tutorial? Consider if this change should be announced on your project blog.
- [ ] Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ] Do all new files have appropriate license header?

# Post Merge Requirements
- [ ] MAINTAINERS: manually trigger the "Publish Package" workflow after merging any PR that indicates `Helm Chart Change`

susanshi commented on pkg/common/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -130,9 +130,23 @@ func (d *azureWIAuthProvider) Provide(ctx context.Context, artifact string) (pro
 	// create registry client and exchange AAD token for registry refresh token
-	refreshTokenClient := containerregistry.NewRefreshTokensClient(serverURL)
+	// TODO: Consider adding authentication client options for multicloud scenarios
+	client, err := azcontainerregistry.NewAuthenticationClient(serverURL, nil) // &AuthenticationClientOptions{ClientOptions: options})
```
suggestion to make a variable ( improved readability for why the param is nil) , assign it to nil then pass into NewAuthenticationClient

susanshi commented:
Hi @shahramk64 , thanks for the PR. please link to the AKS run in your fork once this is ready for review. thanks!

binbin-li commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -135,8 +135,21 @@ func (d *azureManagedIdentityAuthProvider) Provide(ctx context.Context, artifact
-	rt, err := refreshTokenClient.GetFromExchange(ctx, "access_token", artifactHostName, d.tenantID, "", d.identityToken.Token)
+	client, err := azcontainerregistry.NewAuthenticationClient(serverURL, nil) // &AuthenticationClientOptions{ClientOptions: options})
+	if err != nil {
+		return provider.AuthConfig{}, re.ErrorCodeAuthDenied.NewError(re.AuthProvider, "", re.AzureWorkloadIdentityLink, err, "failed to create authentication client for container registry by azure managed identity token", re.HideStackTrace)
```
nit: we can follow the new pattern to generate a Ratify error: https://github.com/ratify-project/ratify/blob/dev/pkg/controllers/utils/verifier.go#L63

shahramk64 replied:
@binbin-li  I tried to mimic the pattern that's already being used in this file (and in azureworkloadidentity.go) in multiple places. Should I go ahead and change all of them?

shahramk64 commented:
@akashsinghal @susanshi 
I added unit tests to azureworkloadidentity_test.go. I would appreciate it if you could have a look at it. Based on what I understood, to mock the exported functions in azcontainerregistry, I needed to use dependency injection, so I refactored  azureworkloadidentity.go for that. 
If this is the right way to write the unit tests, I'll follow the same pattern for azureidentity.go and will try to bring the coverage up to 80%

akashsinghal commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -135,8 +135,21 @@ func (d *azureManagedIdentityAuthProvider) Provide(ctx context.Context, artifact
+	}
+	// refreshTokenClient := containerregistry.NewRefreshTokensClient(serverURL)
+	rt, err := client.ExchangeAADAccessTokenForACRRefreshToken(
+		context.Background(),
```
we should probably use the `ctx` passed into the method instead of using `ctx.Background`

akashsinghal commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -135,8 +135,21 @@ func (d *azureManagedIdentityAuthProvider) Provide(ctx context.Context, artifact
 	// create registry client and exchange AAD token for registry refresh token
-	refreshTokenClient := containerregistry.NewRefreshTokensClient(serverURL)
-	rt, err := refreshTokenClient.GetFromExchange(ctx, "access_token", artifactHostName, d.tenantID, "", d.identityToken.Token)
+	client, err := azcontainerregistry.NewAuthenticationClient(serverURL, nil) // &AuthenticationClientOptions{ClientOptions: options})
```
It seems like the exchange token logic is the same between azureworkloadidentity and azureidentity auth providers. Can we look into introducing a separate helper function?

akashsinghal commented:
> @akashsinghal @susanshi I added unit tests to azureworkloadidentity_test.go. I would appreciate it if you could have a look at it. Based on what I understood, to mock the exported functions in azcontainerregistry, I needed to use dependency injection, so I refactored azureworkloadidentity.go for that. If this is the right way to write the unit tests, I'll follow the same pattern for azureidentity.go and will try to bring the coverage up to 80%

Thanks @shahramk64. I think this mock approach makes sense if the underlying azure client cannot be mocked

akashsinghal commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -29,14 +29,17 @@ import (
+	identityToken           azcore.AccessToken
+	clientID                string
+	tenantID                string
+	authClientFactory       func(serverURL string, options *azcontainerregistry.AuthenticationClientOptions) (AuthClient, error)
```
Where are the default implementation for these methods set for the `MIAuthProvider`?

akashsinghal commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -29,14 +29,17 @@ import (
+	identityToken           azcore.AccessToken
+	clientID                string
+	tenantID                string
+	authClientFactory       func(serverURL string, options *azcontainerregistry.AuthenticationClientOptions) (AuthClient, error)
```
curious, have you considered creating a separate interface for these 3 methods and then having the `MIAuthProvider` type have the new interface instance as a new field? That is essentially the same as what you are doing now except as a separate type. Any preferences @susanshi @duffney @binbin-li ?

susanshi replied:
I see, if the token properties are different type to the grouping of the methods, we should keep them separate.

binbin-li replied:
seems it's mainly for making unit test easier. I would prefer defining interfaces for those methods.

susanshi commented:
Please validate e2e AKS result before merging.

akashsinghal commented on pkg/common/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -21,21 +21,57 @@ import (
+// MetricsReporter defines an interface for reporting metrics.
+type MetricsReporter interface {
+	ReportMetrics(ctx context.Context, duration int64, artifactHostName string)
+}
```
Should we add this in managed identity provider too?

shahramk64 replied:
As discussed, I created an issue to handle this in a separate PR: #1868

akashsinghal commented on pkg/common/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -129,14 +170,29 @@ func (d *azureWIAuthProvider) Provide(ctx context.Context, artifact string) (pro
-	rt, err := refreshTokenClient.GetFromExchange(context.Background(), "access_token", artifactHostName, d.tenantID, "", d.aadToken.AccessToken)
+	response, err := client.ExchangeAADAccessTokenForACRRefreshToken(
+		ctx,
+		azcontainerregistry.PostContentSchemaGrantType("access_token"),
```
add a constant in the helper for `access_token`

akashsinghal commented on pkg/common/oras/authprovider/azure/helper.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+// DefaultAuthClientFactoryImpl is the default implementation of AuthClientFactory.
+type DefaultAuthClientFactoryImpl struct{}
```
nit: re evaluate to see what needs to be exported and what doesn't

akashsinghal commented:
LGTM. Thanks for working on this. Pending AKS e2e validation.

junczhu commented on pkg/common/oras/authprovider/azure/azureworkloadidentity.go:
```diff
@@ -129,14 +170,29 @@ func (d *azureWIAuthProvider) Provide(ctx context.Context, artifact string) (pro
+	var options *azcontainerregistry.AuthenticationClientOptions
+	client, err := d.authClientFactory.CreateAuthClient(serverURL, options)
+	if err != nil {
+		return provider.AuthConfig{}, re.ErrorCodeAuthDenied.NewError(re.AuthProvider, "", re.AzureWorkloadIdentityLink, err, "failed to create authentication client for container registry", re.HideStackTrace)
```
nit: mark another error message as mentioned above

shahramk64 commented:
Here is a link to a successfull e2e AKS run on my forked repo:
https://github.com/shahramk64/forked_ratify/pull/3

binbin-li commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -29,14 +29,44 @@ import (
+	clientID                string
+	tenantID                string
+	authClientFactory       AuthClientFactory
+	getRegistryHost         RegistryHostGetter
```
nit: probably rename to registryHostGetter or registryHostManager?

shahramk64 commented:
Here is a link to the latest e2e AKS run on my forked repo:
https://github.com/shahramk64/forked_ratify/actions/runs/11451048602/job/31860970244?pr=3

susanshi commented on pkg/common/oras/authprovider/azure/azureidentity.go:
```diff
@@ -112,57 +144,59 @@ func (d *azureManagedIdentityAuthProvider) Enabled(_ context.Context) bool {
+	var options *azcontainerregistry.AuthenticationClientOptions
+	client, err := d.authClientFactory.CreateAuthClient(serverURL, options)
+	if err != nil {
+		// return provider.AuthConfig{}, re.ErrorCodeAuthDenied.NewError(re.AuthProvider, "", re.AzureWorkloadIdentityLink, err, "failed to create authentication client for container registry by azure managed identity token", re.HideStackTrace)
```
should this be removed @shahramk64 ?

susanshi reviewed (approved):
thanks for updating the SDK and making significant effort to increase code coverage

binbin-li reviewed (approved):
lgtm, thanks for the sdk upgrade work!
